hera_pspec/pspecdata.py

```diff
@@ -1797,95 +1854,134 @@ def get_analytic_covariance(self, key1, key2, M=None, exact_norm=False, pol=Fals
+            elif model in ['foreground_dependent','outer_product']:
+                E12C21, E12P22, E21starS11, E12starS21, E12P21, E21C12, E21P11, E12starS22 = [0],  [0],  [0],  [0],  [0],  [0],  [0],  [0]
+                E21C11 = np.matmul(np.transpose(E_matrices.conj(), (0,2,1)), C11[time_index])
+                E12C22 = np.matmul(E_matrices, C22[time_index])
+            else:
+                E12C21 = np.matmul(E_matrices, C21[time_index])
+                E12P22 = np.matmul(E_matrices, P22[time_index])
+                E21starS11 = np.matmul(np.transpose(E_matrices, (0,2,1)), S11[time_index])
+                E21C11 = np.matmul(np.transpose(E_matrices.conj(), (0,2,1)), C11[time_index])
+                E12C22 = np.matmul(E_matrices, C22[time_index])
+                E12starS21 = np.matmul(E_matrices.conj(), S21[time_index])
+                E12P21 = np.matmul(E_matrices, P21[time_index])
+                E21C12 = np.matmul(np.transpose(E_matrices.conj(), (0,2,1)), C12[time_index])
+                E21P11 = np.matmul(np.transpose(E_matrices.conj(), (0,2,1)), P11[time_index])
+                E12starS22 = np.matmul(E_matrices.conj(), S22[time_index]) 
             # (spw_Ndlys, spw_Nfreqs, spw_Nfreqs)
 
             # Get q_q, q_qdagger, qdagger_qdagger
-            q_q = np.einsum('bij, cji->bc', E12P22, E21starS11) + np.einsum('bij, cji->bc', E12C21, E12C21)
-            q_qdagger = np.einsum('bij, cji->bc', E12C22, E21C11) + np.einsum('bij, cji->bc', E12P21, E12starS21)
-            qdagger_qdagger = np.einsum('bij, cji->bc', E21C12, E21C12) + np.einsum('bij, cji->bc', E21P11, E12starS22)
+            einstein_path =  np.einsum_path('bij, cji->bc', E12C22, E21C11, optimize='optimal')[0]
+            if np.isclose(E12P22, 0).all() or np.isclose(E21starS11,0).all():
+                q_q = 0.+1.j*0
+            else:
+                q_q = np.einsum('bij, cji->bc', E12P22, E21starS11, optimize=einstein_path)
+            if np.isclose(E12C21, 0).all(): 
+                q_q += 0.+1.j*0
+            else:
+                q_q += np.einsum('bij, cji->bc', E12C21, E12C21, optimize=einstein_path)
+
+            q_qdagger = np.einsum('bij, cji->bc', E12C22, E21C11, optimize=einstein_path) 
+            if np.isclose(E12P21, 0).all() or np.isclose(E12starS21,0).all():
+                q_qdagger += 0.+1.j*0
+            else:
+                q_qdagger += np.einsum('bij, cji->bc', E12P21, E12starS21, optimize=einstein_path)
+            if model == 'foreground_dependent':
+                q_qdagger -= 2*np.einsum('bij, cji->bc', np.multiply(E_matrices, np.diag(C22_autos[time_index])), np.multiply(np.transpose(E_matrices.conj(), (0,2,1)), np.diag(C11_autos[time_index])), optimize=einstein_path)
+                q_qdagger -= np.einsum('bij, cji->bc', np.matmul(E_matrices, C22_outer_product[time_index]), np.matmul(np.transpose(E_matrices.conj(), (0,2,1)), C11_outer_product[time_index]), optimize=einstein_path)
```

Comment:
This doesn't appear to take advantage of the fact that `C11_outer_product` and `C22_outer_product` can be written as the outer product of two vectors. I'm curious---is this because it was actually quicker this way, or is it just that it's more cumbersome to code it in a way that explicitly takes advantage of the outer product? I don't mind either way as long as it's quick enough for our purposes.